### PR TITLE
Refactor komodo_state::NPOINTS

### PR DESCRIPTION
**Pull request by jmjatlanta**
Part of #479

The NPOINTS array required linear searches and never released memory. This PR make NPOINTS a boost::multi_index collection which solves those two issues. In addition, access to the collection is done through member methods of the komodo_state object.

### CODE REVIEW

**Comment by jmjatlanta** [comment]
Note: The method implementations should be moved to komodo_structs.cpp if #476 gets merged in before this one.

**Comment by dimxy** [comment]
I personally would prefer we avoid using boost in favour STL (reasons are STL is better integrated into compiler, better suitable for error-diagnosing and debugging, also not using extra libs reduces learning curve) but this is just a suggestion for consideration

**Comment by dimxy** [comment]
now the TestEvents.komodo_faststateinit_test seems to work okay

**Comment by DeckerSU** [comment]
@jmjatlanta From one side, i highly appreciate your efforts and work on komodod core refactoring, also i like the idea with using `boost::multi_index` container for NPOINTS storage, instead of just an in-memory array. But from other side i worry every time when changes touches notary / notarisations code, bcz it's very important part and we can't allow any, even small errors in it. For review this PR i prepared a kind of separate "test environment" to ensure that new refactored code is fully correct. I transferred old code (namespace `old_space`) and new code (namespace `new_space`) in separate program and tried to ensure that both produce the same results on live data extracted directly from daemon. But in my very small series of tests it turns out that the **logic is broken** somewhere:
```
ize: 8043 (8043, 8043)
[old] komodo_npptr_for_height(2524224): (null)
[new] komodo_npptr(2524224): (null) - [ PASS ]

[old] komodo_npptr_for_height(2524223): (null)
[new] komodo_npptr(2524223): (null) - [ PASS ]

[old] komodo_npptr_for_height(2524210): {"0x0b6154c662432d4a6ba9ba9ee72388b576da4069d5e5aa0f33f0a8952395c895", "0xa5fa719ad82e881342c9bae94c85e3d85cef2808e8bb3cad4d8b37a62afc9624", "0x0", "0x0", 2524223, 2524210, 1, 0, 0, 0, 0}
[new] komodo_npptr(2524210): {"0x0b6154c662432d4a6ba9ba9ee72388b576da4069d5e5aa0f33f0a8952395c895", "0xa5fa719ad82e881342c9bae94c85e3d85cef2808e8bb3cad4d8b37a62afc9624", "0x0", "0x0", 2524223, 2524210, 1, 0, 0, 0, 0} - [ PASS ]

[old] komodo_npptr_for_height(2524209): (null)
[new] komodo_npptr(2524209): (null) - [ PASS ]

[old] komodo_npptr_for_height(2524200): (null)
[new] komodo_npptr(2524200): (null) - [ PASS ]

[old] komodo_npptr_for_height(2524190): {"0x0babd5b8bacec1776f1bf97ea2fefff96e3ca0785def3904a858f3c196569fd3", "0x98207abc680d96d791af3184111da76428cbebd54e2a02fad26f92a872b2c7f0", "0x0", "0x0", 2524204, 2524190, 20, 0, 0, 0, 0}
[new] komodo_npptr(2524190): {"0x0babd5b8bacec1776f1bf97ea2fefff96e3ca0785def3904a858f3c196569fd3", "0x98207abc680d96d791af3184111da76428cbebd54e2a02fad26f92a872b2c7f0", "0x0", "0x0", 2524204, 2524190, 20, 0, 0, 0, 0} - [ PASS ]

[old] komodo_npptr_for_height(2524171): {"0x0babd5b8bacec1776f1bf97ea2fefff96e3ca0785def3904a858f3c196569fd3", "0x98207abc680d96d791af3184111da76428cbebd54e2a02fad26f92a872b2c7f0", "0x0", "0x0", 2524204, 2524190, 20, 0, 0, 0, 0}
[new] komodo_npptr(2524171): (null) - [ FAIL ]

[old] komodo_npptr_for_height(2524170): (null)
[new] komodo_npptr(2524170): (null) - [ PASS ]

[old] komodo_notarizeddata(2524224) = 2524210
[new] komodo_notarizeddata(2524224) = 2524210 - [ PASS ]
[old] komodo_notarizeddata(2524223) = 2524200
[new] komodo_notarizeddata(2524223) = 0 - [ FAIL ]
[old] komodo_notarizeddata(2524210) = 2524190
[new] komodo_notarizeddata(2524210) = 2524190 - [ PASS ]
[old] komodo_notarizeddata(2524209) = 2524190
[new] komodo_notarizeddata(2524209) = 2524190 - [ PASS ]
[old] komodo_notarizeddata(2524200) = 2524180
[new] komodo_notarizeddata(2524200) = 2524180 - [ PASS ]
[old] komodo_notarizeddata(2524192) = 2524170
[new] komodo_notarizeddata(2524192) = 0 - [ FAIL ]
[old] komodo_notarizeddata(2524190) = 2524170
[new] komodo_notarizeddata(2524190) = 2524170 - [ PASS ]
[old] komodo_notarizeddata(2524171) = 2524150
[new] komodo_notarizeddata(2524171) = 2524150 - [ PASS ]
[old] komodo_notarizeddata(2524170) = 2524150
[new] komodo_notarizeddata(2524170) = 2524150 - [ PASS ]
[old] komodo_notarizeddata(2441342) = 2441320
[new] komodo_notarizeddata(2441342) = 0 - [ FAIL ]
```
I mean `komodo_npptr` and `komodo_notarizeddata` new implementation gives wrong results (marked with **[ FAIL ]**) in some cases, different from old, original implementation. And these are the only tests i was able to create, i will no wonder if other hidden errors which is not covered by these tests could exists. 

Thinking this way in trying to save original logic seems we have 3 options:

1. Don't touch the legacy code at all. I don't like this idea, bcz some code indeed needs refactoring, all of these arrays usage instead of STL containers, memsets and strycpys looks ugly. From other side these code is tested in production and at least it works.
2. Completely rewrite legacy code from scratch. This approach is preferable in compare with ours "point surgeries" / "in-place changes", like changing malloc/realloc dynamically created array to STL's vector, but from other side it will require deep understanding of entire idea, kind of "global vision / plan" of all of that processes in head. I mean, whole and detail plan of how notarisations and other things works. 
3. "Compromise mode" in which we will continue "point surgeries" / "in-place changes" for the places which is not related to global / important things like notarisations, and learn the code in parallel to implement the (2) some day.

My test program for #481 is in https://github.com/DeckerSU/events-refactor-test in `fixtry` branch. Sorry for the possible inconveniences, i know that more convenient will be to implement it as a unit test in Google Test Framework, but from other side - compiling small separate program and test some issues in it could be bit faster. Anyway, it's separate from main codebase for now :no_mouth:

**Comment by jmjatlanta** [comment]
@DeckerSU Thank you for your deep efforts in testing. And I agree that we are changing code that affects very sensitive areas. I will take what you have done and work to resolve this issue.

Here are my thoughts and opinions on how to proceed with such refactoring:
1. Break large methods into smaller methods. It is easier to see what should go in and what should come out. 
2. Unit tests will prove it if time is taken to assure edge cases are thought of.
3. Document when the code is unclear - Make it easy for the next guy (the reviewer as well as the guy 4 years from now) to follow.
4. System test in testnet. Put the code through its paces. Test the whole system, but concentrating on the changed areas.

In this case I failed at many of these points. Coincidentally last week I began walking through the notary code to gain a better understanding of how it works. I deployed a testnet with the hope to system test notary changes, among other things.

I agree that not refactoring is not an option. Rewriting from scratch is an option, but often impractical. If you would like a short read, check out [Migrating Legacy Systems](https://www.amazon.com/Migrating-Legacy-Systems-Interfaces-Incremental/dp/1558603301/ref=sr_1_fkmr0_1?crid=1DNQRMKI4AN0N&keywords=migrating+legacy+code&qid=1643111792&sprefix=migrating+legacy+cod%2Caps%2C134&sr=8-1-fkmr0)

I'll get started figuring out what I did wrong and report back. Thanks again for your detailed efforts.

**Comment by jmjatlanta** [comment]
@DeckerSU I have modified the code your tests were running to fix the indexing problems that you found. You may take a look at https://github.com/DeckerSU/events-refactor-test/compare/fixtry...jmjatlanta:fixtry?expand=1 . I also took your tests and turned them into a gtest in this PR. The new tests take time to run due to the old linear search, but it checks every entry and then some extra at the end for bounds checking. It also takes extra time to compile in debug mode due to variable following.

**Comment by jmjatlanta** [comment]
As per the suggestions of @dimxy and @DeckerSU boost::multi_index was removed in favor of std::vector. This may lead to longer search times, but it is thought that the increased complexity is not worth the risk of introducing bugs. In the end, the search times have probably not increased over the original code.

What has changed is that the code uses far less pointers and no malloc/calloc/realloc. In addition, methods that modify the vector have been moved to within the komodo_state object itself.

File: src/komodo_structs.h [hunk 1]
```diff
@@ -53,6 +53,12 @@
 #define KOMODO_ASSETCHAIN_MAXLEN 65
 
 #include "bits256.h"
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/sequenced_index.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/mem_fun.hpp>
+#include <set>
```
Comment (DeckerSU) [inline src/komodo_structs.h:61]: These headers could be deleted now.
```suggestion
```

File: src/komodo_structs.cpp [hunk 2]
```diff
@@ -348,4 +348,175 @@ std::ostream& operator<<(std::ostream& os, const event_pricefeed& in)
...
+ * @note Finding a mach does include other criteria other than height
+ *      such that the checkpoint includes the desired height
+ * @param height the notarized_height desired
+ * @returns the checkpoint or nullptr
+ */
+const notarized_checkpoint *komodo_state::CheckpointAtHeight(int32_t height) const
+{
+    // find the nearest notarization_height
+    // work backwards, get the first one that meets our criteria
+    auto itr = NPOINTS.rbegin();
```
Comment (DeckerSU) [inline src/komodo_structs.cpp:481]:
```suggestion
```
As we have a loop below, this line seems redundant.

File: src/komodo_notary.cpp [hunk 3]
```diff
@@ -283,92 +283,80 @@ int32_t komodo_chosennotary(int32_t *notaryidp,int32_t height,uint8_t *pubkey33,
...
 int32_t komodo_prevMoMheight()
 {
-    static uint256 zero;
-    char symbol[KOMODO_ASSETCHAIN_MAXLEN],dest[KOMODO_ASSETCHAIN_MAXLEN]; int32_t i; struct komodo_state *sp; struct notarized_checkpoint *np = 0;
-    if ( (sp= komodo_stateptr(symbol,dest)) != 0 )
+    char symbol[KOMODO_ASSETCHAIN_MAXLEN];
+    char dest[KOMODO_ASSETCHAIN_MAXLEN];
+    komodo_state *sp;
+
+    if ( (sp= komodo_stateptr(symbol,dest)) != nullptr )
```
Comment (DeckerSU) [inline src/komodo_notary.cpp:316]:
```suggestion
    komodo_state *sp = komodo_stateptr(symbol,dest);
    if ( sp != nullptr )
```
Just to get code more clear.

File: src/komodo_notary.cpp [hunk 4]
```diff
@@ -397,85 +385,61 @@ int32_t komodo_MoMdata(int32_t *notarized_htp,uint256 *MoMp,uint256 *kmdtxidp,in
...
+ * @param[out] notarized_hashp the hash of the notarized block
+ * @param[out] notarized_desttxidp the desttxid
+ * @returns the notarized height
+ */
 int32_t komodo_notarizeddata(int32_t nHeight,uint256 *notarized_hashp,uint256 *notarized_desttxidp)
 {
-    struct notarized_checkpoint *np = 0; int32_t i=0,flag = 0; char symbol[KOMODO_ASSETCHAIN_MAXLEN],dest[KOMODO_ASSETCHAIN_MAXLEN]; struct komodo_state *sp;
-    if ( (sp= komodo_stateptr(symbol,dest)) != 0 )
+    int32_t i=0;
+    bool found = false;
```
Comment (DeckerSU) [inline src/komodo_notary.cpp:401]:
```suggestion
```
Remove as not used.

File: src/komodo_structs.cpp [hunk 5]
```diff
@@ -135,15 +135,15 @@ std::ostream& operator<<(std::ostream& os, const event& in)
  * @param pos the starting position (will advance)
  * @param data_len full length of data
  */
-event_pubkeys::event_pubkeys(uint8_t* data, long& pos, long data_len, int32_t height) : event(EVENT_PUBKEYS, height)
+event_pubkeys::event_pubkeys(uint8_t* data, long& pos, long data_len, int32_t height) : event_pubkeys(height)
```
Comment (DeckerSU) [inline src/komodo_structs.cpp:138]: This change is a part of https://github.com/KomodoPlatform/komodo/pull/510, right? May be we should separate line [#138](https://github.com/KomodoPlatform/komodo/pull/481/files#diff-10bcb1d8edc16456d7b98fedb3834fc3adf81d945d59963596beba41781530b4R138) and line [#146](https://github.com/KomodoPlatform/komodo/pull/481/files#diff-10bcb1d8edc16456d7b98fedb3834fc3adf81d945d59963596beba41781530b4R146) from this PR? Of course, it's important changes, but i would like to consider them in other PR as it's not related to NPOINTS refactoring.

File: src/uint256.h [hunk 6]
```diff
@@ -130,7 +130,7 @@ class uint160 : public base_blob<160> {
  */
 class uint256 : public base_blob<256> {
 public:
-    uint256() {}
+    uint256() : base_blob<256>() {}
```
Comment (DeckerSU) [inline src/uint256.h:133]: Why it's here? The following simple example:
```
#include <iostream>
#include <iomanip>
#include <cstring>

/** Template base class for fixed-sized opaque blobs. */
template<unsigned int BITS>
class base_blob
{
protected:
    // static constexpr int WIDTH = BITS / 8;
    // uint8_t m_data[WIDTH];
    enum { WIDTH=BITS/8 };
    alignas(uint32_t) uint8_t m_data[WIDTH];
public:
    //constexpr base_blob() : m_data() {}
    base_blob()
    {
        std::cout << __FILE__ << "," << __LINE__ << ": " << __func__ << std::endl;
        memset(m_data, 0, sizeof(m_data));
    }
};

class uint256 : public base_blob<256> {
public:
    //constexpr uint256() {}
    uint256() {
        std::cout << __FILE__ << "," << __LINE__ << ": " << __func__ << std::endl;
    }
    void PrintData()
    {
        for (int i = 0; i < WIDTH; ++i) std::cout << std::setfill('0') << std::setw(2) << std::hex << int(m_data[i]) << " ";
        std::cout << std::endl;
    }
    bool IsNull() const
    {
        for (int i = 0; i < WIDTH; i++)
            if (m_data[i] != 0)
                return false;
        return true;
    }
};

int main() {
    uint256 a;
    a.PrintData();

    return 0;
}
```
Shows us that the ctor of `base_blob` is called anyway even without call base class ctor in initializer list, or i miss something? The result of the sample above will be:
```
/home/decker/nullptr.cpp,19: base_blob
/home/decker/nullptr.cpp,28: uint256
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

Comment (jmjatlanta) [reply src/uint256.h:133]: I added that because I thought the base constructor was not being called and garbage was in the array. It turns out the garbage was not caused by what I thought it was. I will back this out.

File: src/komodo_notary.cpp (hunk 4 above)
Comment (jmjatlanta) [reply src/komodo_notary.cpp]: Nor was the int!

**Comment by DeckerSU** [comment]
I almost finished the review of this PR, tested all the changed functions:

- struct notarized_checkpoint *komodo_npptr_for_height(int32_t height, int *idx)
- struct notarized_checkpoint *komodo_npptr(int32_t height)
- int32_t komodo_prevMoMheight()
- int32_t komodo_notarized_height(int32_t *prevMoMheightp,uint256 *hashp,uint256 *txidp)
- int32_t komodo_dpowconfs(int32_t txheight,int32_t numconfs)
- int32_t komodo_MoMdata(int32_t *notarized_htp,uint256 *MoMp,uint256 *kmdtxidp,int32_t height,uint256 *MoMoMp,int32_t *MoMoMoffsetp,int32_t *MoMoMdepthp,int32_t *kmdstartip,int32_t *kmdendip)
- int32_t komodo_notarizeddata(int32_t nHeight,uint256 *notarized_hashp,uint256 *notarized_desttxidp)
- void komodo_notarized_update(struct komodo_state *sp,int32_t nHeight,int32_t notarized_height,uint256 notarized_hash,uint256 notarized_desttxid,uint256 MoM,int32_t MoMdepth)

And corresponding changes in methods of `komodo_state` object. Tests included manual logic check (in compare with original code) and separate auto-tests set for returned values of functions (in compare with old / original implementation) in my own repo. Current version of PR _successfully passed all tests_. The only thing separating me from approval - i want to merge / transfer all of these changes to Komodo-Qt (KomodoOcean) and test few global things, such as "sync from scratch" with monitoring intermediate values of functions, etc. After this will complete i will approve the PR.

Thank you very much, @jmjatlanta for your work and prompt handling of all changes.

**p.s.** I have only two small questions remained related to the code which i want to discuss:

1. I saw the removal of `komodo_rwccdata` functions and all of it's calls. I know that it marked as demo / example, but i don't so much familiar with CC things. So, @dimxy could you confirm that is 100% safe? I'm sure it's safe, but better to have your opinion too.
2. We have `komodo_mutex` lock inside `komodo_notarized_update`, it come from original implementation. It was used to protect NPOINTS array in old impl. and NPOINTS vector in new impl. But new impl. also have `komodo_state::AddCheckpoint` called from `komodo_notarized_update` and `komodo_state::AddCheckpoint` doesn't protected itself. I mean "next guy" could call this method without any protections. 

Also, inside the `komodo_state::AddCheckpoint` we are changing `komodo_state::last`. and as it called from `komodo_notarized_update` it also protected by `komodo_mutex`. But in all other places, like in `SetLastNotarizedXXX` methods we also changing / write the `komodo_state::last` members and these changes doesn't protect with mutex. 

My own opinion - let's leave it as is for now (for this PR may be), but in future should we try to make these places "more thread-safe"?

**Comment by dimxy** [comment]
>  saw the removal of komodo_rwccdata functions and all of it's calls. I know that it marked as demo / example, but i don't so much familiar with CC things. So, @dimxy could you confirm that is 100% safe? I'm sure it's safe, but better to have your opinion too.

We never used this function and as I can see currently it does nothing

**Review by DeckerSU — APPROVED** [review]
**LGTM** now, the only thing - i guess we should reduce size of `test/notarizationdata.tst` to leave there only few data needed for tests executions (some key points, not all values). My offer is leave only data of notarized checkpoints for the heights:
```
{ 2524224, 2524223, 2524210, 2524209, 2524200, 2524190, 2524171, 2524170 }
```
and probably +/- 1 for the each of values.

**Comment by jmjatlanta** [comment]
> i guess we should reduce size of `test/notarizationdata.tst`

I was thinking of this this morning. While the size should be reduced, I am thinking of adding a few more combinations. Overlapping depths was the first thing I thought of. I also would like to work through the MoMoM stuff. I have yet to see where that is used and walk through it.
